### ISSUE TITLE
NLP - [Category] [ near, around, in ] [neighborhood]

### ISSUE THREAD

**Issue by hkrishna** [pelias/pelias#99]
Wouldn't it be cool if we can serve results to the following queries?

```
Restaurants near flatiron
Museums around chelsea
Landmarks in mission district
```

Here's one approach - 

At the time of import if a node has category mapping lets say in case of OSM - ```amenity:restaurant``` and ```cuisine:japanese``` - this means the [given node](https://www.openstreetmap.org/node/3145937665#map=20/40.74408/-73.98945&layers=H) is a japanese restaurant and after a quick admin lookup (part of the import process) we know its neighborhood/locality information. So, in other words we know this is a japanese restaurant in the flatiron neighborhood. 

We could store this information as a shingles field like so - 

```
restaurants flatiron
japanese flatiron
japanese restaurants flatiron
```

At the time of the query we strip out the connector terms like ```near```, ```around```, ```close to```, ```in``` etc (with some sanity checks ofcourse) and add a ```should``` clause to the query - matching the shingles field. When the results are returned, we simply display the ```name.default``` field which contains the name of the japanese restaurant. 

Now, this can be applied to other categories as well such as ```landmarks```, ```airports```, ```museums```, ```churches``` etc like so..

Thoughts?

**Comment by dianashk** [comment]
Why not just search the `category` field? It's also important to keep in mind that people won't use the same word to describe these categories. I might search for `food near flatiron` or `lunch near flatiron` instead of `restaurants new flatiron`

**Comment by mtmail** [comment]
For i18n translations of various terms see http://wiki.openstreetmap.org/wiki/Nominatim/Special_Phrases/FI, http://wiki.openstreetmap.org/wiki/Nominatim/Special_Phrases/FA etc. The Nominatim geocoder crawls the wiki page as it was the easiest solution to allow users to edit terms.

Be prepared for endless requests for more categories. Be it phone booths that take coins or transgender toilets (based on real github issues) or http://wiki.openstreetmap.org/wiki/Tag:vending%3Dexcrement_bags (totally making this up).

**Comment by missinglink** [comment]
lol @mtmail 

I like the idea of providing autocomplete 'hints' of venues near you, in particular for restaurants and cuisine (for which we have the categories already set up!)

here are some screenshots from another geocoder you may be familiar with:
![](http://missinglink.embed.s3.amazonaws.com/g-near1.png)
![](http://missinglink.embed.s3.amazonaws.com/g-near2.png)

foursquare/twofishes are all about the what/where and so they are also a good one to have a look at.

**Comment by hkrishna** [comment]
@dianashk yes, first step is for our API to have an endpoint wherein you can search categories straight up.. https://github.com/pelias/api/issues/146

@mtmail thanks for the links - those are really helpful. And please let us know if we are missing anything - your feedback/suggestions are valuable :)

@missinglink I like the idea of providing autocomplete 'hints' of venues nearby as well. I've created this issue #100 - do add more details there if necessary.

**Comment by michaelkirk** [comment]
I hope this is still something the project is interested in incorporating. I want this to power a search field like in the screenshot, and have spent part of today digging in to pelias to see how it might work:

<img src=https://camo.githubusercontent.com/764a8230aea1ad3dd225c0485082f78566cef5fcd1913bd5fcf6963bfee3bca6/687474703a2f2f6d697373696e676c696e6b2e656d6265642e73332e616d617a6f6e6177732e636f6d2f672d6e656172312e706e67 />

I'd want something like this for both /search and /autocomplete (#100). If it were just me, I'd bake this behavior into the existing /search and /autocomplete endpoint, because I always want to include good categorical matches. I actually had assumed it worked that way when I confusedly opened #938,  but maybe that's too broad of an assumption - should this be a different set of endpoints, or some kind of configuration?

My simple POC was just to make a `multi_match` query across both `name` and `category` fields. Probably there's some subtlety to how they should be scored, but I haven't gotten there.

With my current WIP, you need to search for the literal category like `food:coffee_shop`. Obviously the user will be searching for something more like "coffee shop" or maybe just "coffee". I'm an elasticsearch neophyte, and not really sure how to model this. If I were to blunder onward, I think I'd start with something like elasticsearch's [synonyms](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html), so that "coffee", "coffee shop", and "cafe" could be mapped to the "food::coffee_shop" category.

I'm really not sure if that's the right way to go about it though. Currently categories are "keywords" but I think I want to be able to match them more like text.

Anyway, just throwing out some bad ideas in hopes someone with better ideas might respond. 😅

**Comment by CeallachKN** [comment]
Just my two cents in here.

1. I think this shouldn't alter the standard behaviour or be set in config, rather a true/false/"use these categorys"-parameter for the search/autocomplete-query as not every use case would need this funcionality.
3. The custom-importers would need to be considered, since there can be endless possibilitys for naming the categorys. I think these would need to be imported as "alt-categoryname" since detecting synonyms on these, besides splitting multiword phrases, would be near impossible to achive
